src/planner/partialize.c

```diff
@@ -185,6 +197,657 @@ partialize_agg_paths(RelOptInfo *rel)
+	bool can_hash = grouping_is_hashable(parse->groupClause) &&
+					!ts_is_gapfill_path(linitial(output_rel->pathlist)) && enable_hashagg;
+
+	Assert(extra != NULL);
+	GroupPathExtraData *extra_data = (GroupPathExtraData *) extra;
+
+	/* Determine the number of groups from the already planned aggregation */
+	AggPath *existing_agg_path = get_existing_agg_path(output_rel);
+	if (existing_agg_path == NULL)
+		return;
+
+	/* Skip partial aggregations already created by _timescaledb_functions.partialize_agg */
+	if (existing_agg_path->aggsplit == AGGSPLIT_INITIAL_SERIAL)
+		return;
+
+/* Don't replan aggregation if it contains already partials or non-serializable aggregates */
+#if PG14_LT
+	AggClauseCosts agg_costs;
+	MemSet(&agg_costs, 0, sizeof(AggClauseCosts));
+	get_agg_clause_costs_compat(root, (Node *) root->processed_tlist, AGGSPLIT_SIMPLE, &agg_costs);
+	get_agg_clause_costs_compat(root, parse->havingQual, AGGSPLIT_SIMPLE, &agg_costs);
+
+	if (agg_costs.hasNonPartial || agg_costs.hasNonSerial)
+#else
+	if (root->hasNonPartialAggs || root->hasNonSerialAggs)
+#endif
+		return;
+
+	double d_num_groups = existing_agg_path->numGroups;
+	Assert(d_num_groups > 0);
+
+	/* Construct partial group agg upper relation */
+	RelOptInfo *partially_grouped_rel =
+		fetch_upper_rel(root, UPPERREL_PARTIAL_GROUP_AGG, input_rel->relids);
+	partially_grouped_rel->consider_parallel = input_rel->consider_parallel;
+	partially_grouped_rel->reloptkind = input_rel->reloptkind;
+	partially_grouped_rel->serverid = input_rel->serverid;
+	partially_grouped_rel->userid = input_rel->userid;
+	partially_grouped_rel->useridiscurrent = input_rel->useridiscurrent;
+	partially_grouped_rel->fdwroutine = input_rel->fdwroutine;
+
+	/* Build target list for partial aggregate paths */
+	PathTarget *grouping_target = output_rel->reltarget;
+	PathTarget *partial_grouping_target = ts_make_partial_grouping_target(root, grouping_target);
+	partially_grouped_rel->reltarget = partial_grouping_target;
+
+	/* Calculate aggregation costs */
+	if (!extra_data->partial_costs_set)
+	{
+		/* partial phase */
+		get_agg_clause_costs_compat(root,
+									(Node *) partial_grouping_target->exprs,
+									AGGSPLIT_INITIAL_SERIAL,
+									&extra_data->agg_partial_costs);
+
+		/* final phase */
+		get_agg_clause_costs_compat(root,
+									(Node *) root->upper_targets[UPPERREL_GROUP_AGG]->exprs,
+									AGGSPLIT_FINAL_DESERIAL,
+									&extra_data->agg_final_costs);
+
+		extra_data->partial_costs_set = true;
+	}
+
+	/* Generate the aggregation pushdown path */
+	Path *cheapest_total_path = get_best_total_path(input_rel);
+	Assert(cheapest_total_path != NULL);
+	generate_agg_pushdown_path(root,
+							   cheapest_total_path,
+							   output_rel,
+							   partially_grouped_rel,
+							   grouping_target,
+							   partial_grouping_target,
+							   can_sort,
+							   can_hash,
+							   d_num_groups,
+							   extra_data);
+
+	/* The same as above but for partial paths */
```

Comment:
I find this confusing: We just generated push-down paths above, which in my understanding are also partial paths. (Please correct me if I am wrong.) But here it sounds like the partial paths are generated here, implying it was not done in the previous step.

I think what is happening is that we are conflating partial and parallel. (And maybe PostgreSQL does too.)

So, in my understanding, we can have partial paths that are both parallel and non-parallel. Thus:

1. We use non-parallel partials to push down aggregates to individual chunk relations ("partitionwise" agg).
2. We can also generate parallel execution paths, which also require partials since they execute in different workers ("parallel" agg).

Thus, I think in the previous step we generated partial push-down paths that are non-parallel. Here we generate partial paths to be executed in parallel (given the `consider_parallel` check).

Unless I am wrong, I think PostgreSQL uses `rel->pathlist` to store non-parallel paths and `rel->partial_pathlist` to store parallel paths although both can contain partials if we do partitionwise aggregation (which makes the naming super-confusing). I guess PostgreSQL's naming convention here is confusing and legacy and probably a result of initially only using partial paths in the parallel case.


Currently, I think it is hard to understand the difference between what `generate_agg_pushdown_path` does above compared to `generate_partial_agg_pushdown_path` below. And I am not even sure my understanding is correct here. Is there a way to clarify in function names and comments?


I am wondering if it would not be more clear to say that above we generate "partitionwise" paths, while here we generate the corresponding "parallel" paths?